ipa-core/src/protocol/context/mod.rs

```diff
@@ -344,6 +374,154 @@ impl<'a> Inner<'a> {
     }
 }
 
+/// Reshards the given set of shares across all shards available on the current helper. It uses the
+/// provided function to determine destination shards for each input row.
+///
+/// The resulting vector contains all shares sent to this shard by other shards and shares selected
+/// locally by `shard_picker` fn to stay on this shard.
+///
+/// Resharding opens one communication channel per peer (N^2 channels will be open in total,
+/// N per shard). Each channel stays open until the very last row is processed, then they are explicitly
+/// closed, even if nothing has been communicated between that pair.
+///
+/// ## Shard picking considerations
+/// It is expected for `shard_picker` to select shards uniformly, by either using [`prss`] or sampling
+/// random values with enough entropy. Failure to do so may lead to extra memory overhead - this
+/// function uses the conservative `1.2` coefficent to estimate the number of records per shard after
+/// resharding is completed, according to our [`calculations`] that is sufficient with 2^-60 failure
+/// probability. This is a very conservative estimate, assuming 1M events per shard and 100k shards.
+///
+/// [`calculations`]: https://docs.google.com/document/d/1vej6tYgNV3GWcldD4tl7a4Z9EeZwda3F5u7roPGArlU/
+///
+/// ## Panics
+/// When `shard_picker` returns an out-of-bounds index.
+///
+/// ## Errors
+/// If cross-shard communication fails
```

Comment:
yea we don't have retry mechanisms built-in at the application layer right now and we rely on transport layer (TCP) to provide reliable channels. We may build something for that purpose if we see that TCP does not satisfy our needs, but it shouldn't be visible to MPC because there isn't anything you can do here that you can't do at the infrastructure layer.